drivers: regulator-fixed: extend api of driver (list/count voltages)

xakep-amatop opened the pull request:
Allow to create fixed-regulator nodes without gpio for en/dis power source.
Allow properties `regulator-min-microvolt` and `regulator-max-microvolt` for fixed regulator, note: they should be equal.
Add simple functions for getting list of allowed and count of voltages.

xakep-amatop committed:
drivers: regulator-fixed: add possibility to work without enable pins

Possible situation is that in some driver, devices can be controlled in
different ways: in some, we can only turn the power on or off, in others,
we can only control the voltage, and in some, we can control power supply
or voltage level. There may also be devices where there is no control
over power supply at all. A clear example of this can be eMMC devices
where the voltage is usually fixed and they are always powered on.
However, we would like to have a common code for controlling all the
mentioned types of devices, at least the driver shouldn't worry about the
implementation details of voltage regulators. Therefore, there may exist
empty regulators - regulators that only contain information about the
supported voltage, and we cannot change anything in them. The device tree
node description for such a regulator is only necessary for compatibility
with other regulators. Hence, we need to add the possibility of the
existence of such a dummy fixed-regulator.

In this commit, support for a fixed dummy regulator without the ability
for any control has been added. Note that such support also exists in the
Linux kernel. In other words, the logic of the fixed regulator has been
aligned with the logic of the fixed regulator inside the Linux kernel.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

gmarull commented on drivers/regulator/regulator_fixed.c:
```diff
@@ -63,38 +96,46 @@ static int regulator_fixed_init(const struct device *dev)
-		ret = gpio_pin_configure_dt(&cfg->enable, GPIO_OUTPUT_ACTIVE);
-		if (ret < 0) {
-			return ret;
+	if (cfg->enable.port) {
```
```suggestion
	if (cfg->enable.port != NULL) {
```

gmarull commented on drivers/regulator/regulator_fixed.c:
```diff
@@ -63,38 +96,46 @@ static int regulator_fixed_init(const struct device *dev)
 	return regulator_common_init(dev, init_enabled);
 }
 
+#define REGULATOR_REQUIRE_FIXED_VOLTAGE(node_id)                               \
+	BUILD_ASSERT(DT_PROP_OR(node_id, regulator_min_microvolt, 0) ==        \
+		     DT_PROP_OR(node_id, regulator_max_microvolt, 0),          \
+		     "Regulator requires fixed voltages")
+
 #define REGULATOR_FIXED_DEFINE(inst)                                           \
+	REGULATOR_REQUIRE_FIXED_VOLTAGE(DT_DRV_INST(inst));                    \
```
nit: build assert can be moved below, it's not shared for anything else here.

gmarull commented on include/zephyr/drivers/regulator.h:
```diff
@@ -567,6 +567,21 @@ static inline int regulator_get_error_flags(const struct device *dev,
 	return api->get_error_flags(dev, flags);
 }
 
+/**
+ * @brief Get minimal voltage from common config.
+ *
+ * @param dev Regulator device instance.
+ *
+ * @retval minimal supported voltage from common config.
+ */
+static inline int32_t regulator_get_cfg_min_uv(const struct device *dev)
+{
+	const struct regulator_common_config *cmn =
+		(const struct regulator_common_config *)dev->config;
+
+	return cmn->min_uv;
+}
+
```
this is an internal API, meant for drivers only. Please name it `regular_common_get_min_voltage()` and move it to the internal section where common functions are present.

gmarull commented on drivers/regulator/regulator_fixed.c:
```diff
@@ -47,12 +51,41 @@ static int regulator_fixed_disable(const struct device *dev)
+					unsigned int idx,
+					int32_t *volt_uv)
+{
+	int32_t min_uv = regulator_get_cfg_min_uv(dev);
+
+	if (idx != 0) {
+		return -EINVAL;
+	}
+
```
nit: assignment can happen after index check.

gmarull commented on drivers/regulator/regulator_fixed.c:
```diff
@@ -31,6 +31,10 @@ static int regulator_fixed_enable(const struct device *dev)
 	const struct regulator_fixed_config *cfg = dev->config;
 	int ret;
 
+	if (!cfg->enable.port) {
+		return -ENOTSUP;
+	}
+
```
what is the point of a fixed regulator that can't be controlled? Also, this seems to deviate from Linux where either clocks/gpio/power domain are required.

xakep-amatop replied:
A driver can request the current voltage of regulator, and for such drivers, we can utilize distinct regulators for various nodes, for example, fixed w/o control for one node of the driver and gpio regulator for another node

Many fixed regulators do not require clocks, GPIO, or power domains within the Linux kernel. For instance, refer to: https://elixir.bootlin.com/linux/latest/source/arch/arm64/boot/dts/renesas/salvator-common.dtsi#L181

gmarull replied:
ok, fair enough. Doxygen should also be updated to mention this is a valid case:

```
@retval -ENOTSUP If regulator enablement can not be controlled
```

xakep-amatop replied:
> Doxygen should also be updated to mention this is a valid case

is it really needed? as far as I see, we have doxygen style comment only for common enable function, `regulator_enable`:
```
 * @retval 0 If regulator has been successfully enabled.
 * @retval -errno Negative errno in case of failure.
 */
int regulator_enable(const struct device *dev);
```

For some drivers, this could have another meaning. Furthermore, this function (`regulator_fixed_enable`) won't be called in cases where we have the `always-on` property in the device tree (`dts`), see `regulator_enable`:
```
	/* regulator must stay always on */
	if  ((config->flags & REGULATOR_ALWAYS_ON) != 0U) {
		return 0;
	}
```

I believe that `always-on` and `enable-gpios` should indeed be mutually exclusive for the fixed regulator driver, and it would be beneficial to add a build assert:
```
BUILD_ASSERT(DT_PROP_HAS_IDX(DT_DRV_INST(inst), enable_gpios, 0) ^       \
             DT_PROP_OR(DT_DRV_INST(inst), regulator_always_on, 0),      \
             "always-on and enable-gpios should be mutually exclusive"); \
```

What are your thoughts on this?

gmarull replied:
> > Doxygen should also be updated to mention this is a valid case
> 
> is it really needed? as far as I see, we have doxygen style comment only for common enable function, `regulator_enable`:
> 
> ```
>  * @retval 0 If regulator has been successfully enabled.
>  * @retval -errno Negative errno in case of failure.
>  */
> int regulator_enable(const struct device *dev);
> ```

`-errno` case is really reserved for propagated errors (e.g. from a failed transfer). The `-ENOTSUP` case here is not strictly a failure.

> 
> For some drivers, this could have another meaning. Furthermore, this function (`regulator_fixed_enable`) won't be called in cases where we have the `always-on` property in the device tree (`dts`), see `regulator_enable`:
> 
> ```
> 	/* regulator must stay always on */
> 	if  ((config->flags & REGULATOR_ALWAYS_ON) != 0U) {
> 		return 0;
> 	}
> ```
> 

This case is differentiated because it returns 0. Actually, always-on/boot-on don't make much sense in a fixed regulator without gpios.

> I believe that `always-on` and `enable-gpios` should indeed be mutually exclusive for the fixed regulator driver, and it would be beneficial to add a build assert:
> 
> ```
> BUILD_ASSERT(DT_PROP_HAS_IDX(DT_DRV_INST(inst), enable_gpios, 0) ^       \
>              DT_PROP_OR(DT_DRV_INST(inst), regulator_always_on, 0),      \
>              "always-on and enable-gpios should be mutually exclusive"); \
> ```
> 
> What are your thoughts on this?

I think it makes sense, but right now, it's used to turn on the regulator during init time (set to active forever). Also, I think we're using the `regulator-boot-on` incorrectly, as it should be used to flag that the regulator is already on (from a previous stage).

xakep-amatop replied:
> it's used to turn on the regulator during init time (set to active forever). 

I see, than there is no sense to add a build assert,
I have updated `regulator_enable` description

gmarull commented on include/zephyr/drivers/regulator.h:
```diff
@@ -242,6 +243,21 @@ static inline bool regulator_common_is_init_enabled(const struct device *dev)
 	return (config->flags & REGULATOR_INIT_ENABLED) != 0U;
 }
 
+/**
+ * @brief Get minimal voltage from common config.
+ *
+ * @param dev Regulator device instance.
+ *
+ * @retval minimal supported voltage from common config.
+ */
+static inline int32_t regulator_get_cfg_min_uv(const struct device *dev)
+{
+	const struct regulator_common_config *cmn =
+		(const struct regulator_common_config *)dev->config;
+
+	return cmn->min_uv;
+}
```
```suggestion
/**
 * @brief Get minimum supported voltage.
 *
 * @param dev Regulator device instance.
 * @param min_uv Where minimum voltage will be stored, in microvolts.
 *
 * @retval 0 If successful
 * @retval -ENOENT If minimum voltage is not specified.
 */
static inline int regulator_common_get_min_voltage(const struct device *dev, int32_t *min_uv)
{
	const struct regulator_common_config *config =
		(const struct regulator_common_config *)dev->config;

	if (config->min_uv == INT32_MIN) {
		return -ENOENT;
	}

	*min_uv = config->min_uv;
	
	return 0;
}
```

This avoids exposing common layer internals to drivers (ie INT32_MIN).

gmarull reviewed (changes requested):
please split changes in 2 commits, one making enable optional (explaining the reasons) and a second adding support for fixed voltage reporting

xakep-amatop committed:
drivers: regulator-fixed: extend api of driver (list/count voltages)

Allow properties 'regulator-min-microvolt' and 'regulator-max-microvolt'
for fixed regulators: Note: they should be equal.

Add simple functions for getting list of allowed and count of voltages.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop commented:
> please split changes in 2 commits, one making enable optional (explaining the reasons) and a second adding support for fixed voltage reporting

done

gmarull commented on include/zephyr/drivers/regulator.h:
```diff
@@ -322,6 +322,7 @@ static inline int regulator_parent_ship_mode(const struct device *dev)
  *
  * @retval 0 If regulator has been successfully enabled.
  * @retval -errno Negative errno in case of failure.
+ * @retval -ENOTSUP If regulator enablement can not be controlled.
  */
 int regulator_enable(const struct device *dev);
```
nit: do the same for disable call

gmarull commented on dts/bindings/regulator/regulator-fixed.yaml:
```diff
@@ -1,7 +1,8 @@
 # Copyright 2019-2020 Peter Bigot Consulting, LLC
+# Copyright 2023 EPAM Systems
```
nit: seems excessive for one line change

xakep-amatop replied:
in the next commit I added more changes

gmarull commented on drivers/regulator/regulator_fixed.c:
```diff
@@ -59,9 +59,33 @@ static int regulator_fixed_disable(const struct device *dev)
 	return gpio_pin_set_dt(&cfg->enable, 0);
 }
 
+static unsigned int regulator_fixed_count_voltages(const struct device *dev)
+{
+	int32_t min_uv;
+
+	return !regulator_common_get_min_voltage(dev, &min_uv);
```
function retuns an `unsigned int`, not a boolean. Also, `!` operation is done on a non-boolean result.

```
return (regulator_common_get_min_voltage(dev, &min_uv) < 0) ? 0U : 1U;
```

xakep-amatop replied:
> function retuns an unsigned int, not a boolean. 

I see no problems here, function returns 0 or 1 after `!`

> `!` operation is done on a non-boolean result.

no problems here too, pls read C standard: 
```
6.5.3.3 Unary arithmetic operators
Constraints
1 The operand of the unary + or - operator shall have arithmetic type; of the ~ operator,
integer type; of the ! operator, scalar type.
```
```
21 Arithmetic types and pointer types are collectively called scalar types.
```
there is any restriction that operand should be boolean

gmarull replied:
we're following most MISRA rules (even if not applied consistently)

xakep-amatop replied:
Could you please point out the exact MISRA rule that disallows this line?

gmarull replied:
Rule 14.4 https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html

xakep-amatop replied:
sorry, but it is not neither `an if statement`  nor `an iteration-statement`

MISRA doesn't give a definition of this terms, but C standard does and from standard it is clear that `if` and `an iteration` statements should contain special keywords
pls, read from C99 standard the next chapters:
* 6.8.4 Selection statements
* 6.8.5 Iteration statements

gmarull replied:
Please check rules 10.1/3, I got confused by https://github.com/zephyrproject-rtos/zephyr/pull/61240#discussion_r1289891457

xakep-amatop replied:
ok, I read these rules carefully and agree with you that this line breaks some of MISRA rules
thank you for explanation

gmarull commented on drivers/regulator/regulator_fixed.c:
```diff
@@ -59,9 +59,33 @@ static int regulator_fixed_disable(const struct device *dev)
+		return -EINVAL;
+	}
+
+	if (regulator_common_get_min_voltage(dev, volt_uv)) {
```
`if (regulator_common_get_min_voltage(dev, volt_uv) < 0)`

xakep-amatop replied:
I see no reason to do it, only `0` ret value means successful execution of function according description

gmarull replied:
ditto

gmarull commented on include/zephyr/drivers/regulator.h:
```diff
@@ -349,6 +373,7 @@ bool regulator_is_enabled(const struct device *dev);
  *
  * @retval 0 If regulator has been successfully disabled.
  * @retval -errno Negative errno in case of failure.
+ * @retval -ENOTSUP If regulator can not be disabled.
```
```suggestion
* @retval -ENOTSUP If regulator disablement can not be controlled.
```

just to be in line with enable?

xakep-amatop replied:
changed

xakep-amatop commented:
@gmarull thank you for the review
